### PR TITLE
Migrate examples project to Gradle subproject

### PR DESCRIPTION
**Pull request by kiview**
I'm not touching the docs here, this was just for migrating the existing Maven example project.
Let's see if it works on CircleCI also...

I thought it would be cute to have a sub-subproject structure, like `:examples:spring-boot`.
First I wanted to do it for our modules as well, but it would break the existing cross subproject dependencies, so I left it as is for modules.

### CODE REVIEW

File: examples/spring-boot/build.gradle [hunk 1]
```diff
@@ -0,0 +1,23 @@
+buildscript {
```
Comment (bsideup) [inline examples/spring-boot/build.gradle]: just:
```groovy
plugins {
	id 'org.springframework.boot' version '2.1.2.RELEASE'
}
```

**Comment by bsideup** [comment]
@kiview I would really recommend using `includeBuild`.
Otherwise, our examples are not copyable (`testCompile project(":kafka")` instead of `testCompile 'org.testcontainers:kafka'`)

**Comment by kiview** [comment]
I wasn't aware of `includeBuild`, sounds good.

File: settings.gradle [hunk 2]
```diff
@@ -37,5 +37,14 @@ file('modules').eachDir { dir ->
     project(":${dir.name}").projectDir = dir
 }
 
+file('examples').eachDir { dir ->
+    if (dir.name == 'build') {
+        return
+    }
+
+    include "examples:${dir.name}"
+    project(":examples:${dir.name}").projectDir = dir
```
Comment (bmuschko) [inline settings.gradle]: Treating each example as subproject might become a little issue for scalability (Gradle's configuration time) in the long run. Imagine you have hundreds of example projects here. Probably won't happen very soon but is something to consider. I'd use TestKit here and execute them in parallel to speed up the build.

Happy to send a PR once you merged this one.

File: examples/spring-boot/build.gradle [hunk 3]
```diff
@@ -0,0 +1,23 @@
...
+apply plugin: 'org.springframework.boot'
+apply plugin: 'io.spring.dependency-management'
+
+dependencies {
+    implementation 'org.springframework.boot:spring-boot-starter-data-jpa'
+    implementation 'org.springframework.boot:spring-boot-starter-data-redis'
+    implementation 'org.springframework.boot:spring-boot-starter-web'
+    runtimeOnly 'org.postgresql:postgresql'
+    testImplementation 'org.springframework.boot:spring-boot-starter-test'
+    testCompile project(':postgresql')
```
Comment (bmuschko) [inline examples/spring-boot/build.gradle]: Users won't be able to execute the sample as-is. This project relies on other modules of the build. Optimally, you'd inject the TestContainer code "under test" at runtime with TestKit. I'd have to try this out myself but I think it's possible.

Comment (bmuschko) [inline examples/spring-boot/build.gradle]: This configuration should be `testImplementation` to align them with the other configurations used above.

File: examples/selenium-container/src/test/java/SeleniumContainerTest.java [hunk 4]
```diff
@@ -36,7 +36,7 @@
     @Rule
     public BrowserWebDriverContainer chrome = new BrowserWebDriverContainer()
                                                     .withCapabilities(new ChromeOptions())
-                                                    .withRecordingMode(RECORD_ALL, new File("target"));
+                                                    .withRecordingMode(RECORD_ALL, new File("build"));
```
Comment (bmuschko) [inline examples/selenium-container/src/test/java/SeleniumContainerTest.java]: Does this potentially break the Maven example?

File: examples/redis-backed-cache-testng/build.gradle [hunk 5]
```diff
@@ -0,0 +1,13 @@
+dependencies {
+    compile 'redis.clients:jedis:3.0.1'
```
Comment (bmuschko) [inline examples/redis-backed-cache-testng/build.gradle]: Should use the "new" Gradle configurations of the Java plugin e.g. `implementation` etc. Same for all the other example projects.

File: examples/selenium-container/src/test/java/SeleniumContainerTest.java (hunk 4 above)
Comment (kiview) [reply examples/selenium-container/src/test/java/SeleniumContainerTest.java]: It does, but the idea is to remove the Maven examples, since we really don't want to maintain both build systems for the examples.

File: examples/redis-backed-cache-testng/build.gradle (hunk 5 above)
Comment (kiview) [reply examples/redis-backed-cache-testng/build.gradle]: If I recall correctly, @bsideup had some kind of antipathy against them :smile: 
But yes, we should have a consistent style and going further the new configs are the way to go I assume.

File: settings.gradle (hunk 2 above)
Comment (kiview) [reply settings.gradle]: That's a great idea! I'll try to get this merged soon without too much changes so we have finally everything on Gradle and then we can work on improving it :slightly_smiling_face:

Comment (bsideup) [reply settings.gradle]: The composite builds should solve this as well :)

File: examples/spring-boot/build.gradle (hunk 1 above)
Comment (kiview) [reply examples/spring-boot/build.gradle]: That's what https://start.spring.io/ generates by the way :wink:

Comment (bmuschko) [reply examples/spring-boot/build.gradle]: You may want o open an issue for the Spring Start project. The plugins DSL is the preferred syntax.

File: examples/selenium-container/src/test/java/SeleniumContainerTest.java (hunk 4 above)
Comment (bmuschko) [reply examples/selenium-container/src/test/java/SeleniumContainerTest.java]: I see. Thanks for clarifying.

File: examples/redis-backed-cache-testng/build.gradle (hunk 5 above)
Comment (bmuschko) [reply examples/redis-backed-cache-testng/build.gradle]: The "old" configurations like `compile` and friends have been soft-deprecated. IMHO Gradle users don't know what that means but essentially this means that the "new" ones should be preferred. The plan is for the old ones to go away.

File: examples/spring-boot/build.gradle (hunk 3 above)
Comment (bmuschko) [reply examples/spring-boot/build.gradle]: I think that's a good first step. I see the following benefits:

- Developers of this project can use auto-completion as each example project is part of the multi-project builds.
- Users of TestContainers can copy/paste the example code and it should work as-is.

There's one thing I'd personally be missing:

- You can really deeply test example code from a user's perspective. I think it would be good if you could run an example project as a user would and then verify a specific outcome (e.g. tests finish successfully or the test console output should contain x).

Overall, I am happy with your approach. It will unblock other documentation effort. BTW: Do your changes also use the same approach for the quickstart examples?

Comment (bsideup) [reply examples/spring-boot/build.gradle]: > * I think it would be good if you could run an example project as a user would and then verify a specific outcome (e.g. tests finish successfully or the test console output should contain x).

We do run them as a black box (think `./gradlew test`). More specific tests (e.g. for the standard output, or env variables / properties) **IMO** should be done as part of the integration testing suite of the core module

> BTW: Do your changes also use the same approach for the quickstart examples?

We plan to merge both under the same "examples" set of projects

File: examples/spring-boot/build.gradle (hunk 1 above)
Comment (bmuschko) [reply examples/spring-boot/build.gradle]: Looks like the issue already exists: https://github.com/spring-io/initializr/issues/692

File: examples/selenium-container/build.gradle [hunk 6]
```diff
@@ -0,0 +1,16 @@
+plugins {
```
Comment (bsideup) [inline examples/selenium-container/build.gradle]: Spring Boot project, not sure how it even worked without the plugin :D

**Comment by bmuschko** [comment]
@kiview Do you have a target date for getting this PR merged? Would love to contribute more examples.

**Comment by kiview** [comment]
I'd be good to merge, let's have @bsideup a final look :slightly_smiling_face:

File: .dependabot/config.yml [hunk 7]
```diff
@@ -80,9 +80,6 @@ update_configs:
     update_schedule: "daily"
 
 # Examples
-  - package_manager: "java:maven"
+  - package_manager: "java:gradle"
```
Comment (bsideup) [inline .dependabot/config.yml]: why did you remove `automerged_updates`? :)

File: .travis.yml [hunk 8]
```diff
@@ -31,6 +31,9 @@ jobs:
     - env: [ NAME=modules ]
       script: ./gradlew check -x testcontainers:check -x selenium:check --scan --no-daemon
 
+    - env: [ NAME=examples ]
```
Comment (bsideup) [inline .travis.yml]: AFAIR we were running examples only on CircleCI for speed reasons

File: examples/disque-job-queue/build.gradle [hunk 9]
```diff
@@ -0,0 +1,18 @@
+plugins {
+    id 'java'
+}
+
+repositories {
+    jcenter()
+}
+
+dependencies {
+    compileOnly 'org.slf4j:slf4j-api:1.7.25'
```
Comment (bsideup) [inline examples/disque-job-queue/build.gradle]: why is it `compileOnly`?

File: examples/singleton-container/build.gradle [hunk 10]
```diff
@@ -0,0 +1,18 @@
...
+}
+
+dependencies {
+
+    implementation 'redis.clients:jedis:3.0.1'
+    implementation 'com.google.code.gson:gson:2.8.5'
+    implementation 'com.google.guava:guava:23.0'
+    compileOnly 'org.slf4j:slf4j-api:1.7.25'
+
+    testCompile 'ch.qos.logback:logback-classic:1.2.3'
```
Comment (bsideup) [inline examples/singleton-container/build.gradle]:
```suggestion
    testImplementation 'ch.qos.logback:logback-classic:1.2.3'
```

File: examples/singleton-container/src/test/java/com/example/AbstractIntegrationTest.java [hunk 11]
```diff
@@ -4,10 +4,10 @@
...
 public abstract class AbstractIntegrationTest {
 
-	public static final GenericContainer redis = new GenericContainer("redis:3.0.6")
-                                            .withExposedPorts(6379);
-	
-	static {
-		redis.start();
-	}
+    public static final GenericContainer redis = new GenericContainer("redis:3.0.6")
+            .withExposedPorts(6379);
```
Comment (bsideup) [inline examples/singleton-container/src/test/java/com/example/AbstractIntegrationTest.java]: is it IntelliJ's formatting? o_O

Comment (kiview) [reply examples/singleton-container/src/test/java/com/example/AbstractIntegrationTest.java]: Yes, or our `.editorconfig`?

File: .travis.yml (hunk 8 above)
Comment (kiview) [reply .travis.yml]: Examples on Travis with Maven was PITA because of snapshot dependency and strange race conditions.

I'd argue, we can risk to enable them now with composite build.

File: .dependabot/config.yml (hunk 7 above)
Comment (kiview) [reply .dependabot/config.yml]: whoooooops

File: examples/disque-job-queue/build.gradle (hunk 9 above)
Comment (kiview) [reply examples/disque-job-queue/build.gradle]: It was `provided` in the Maven pom.xml.
Let's just remove this.

**Comment by kiview** [comment]
@bsideup Are we good to merge?

**Comment by bsideup** [comment]
@kiview yes 🚀

**Comment by kiview** [comment]
@bmuschko We just merged the PR, we'd love to get more examples if you have some good ideas :slightly_smiling_face:

**Comment by bmuschko** [comment]
@kiview Nice, I will check it out in the coming days.

**Comment by rnorth** [comment]
Released in https://github.com/testcontainers/testcontainers-java/releases/tag/1.10.7!
